### ISSUE TITLE
Support task arrays (horizontally scaled)

### ISSUE THREAD

**Issue by jbingham** [ohsu-comp-bio/funnel#41]
Popular job schedulers like SGE, Slurm, and HTCondor support a single submission of an array of like tasks to be executed in parallel that differ only in their input parameters. Eg, an array of like tasks that each run on a different input file. Or a parameter sweep across all values of a variable from 1 to 100. 

Of course, users can do this today with a loop over the input values. But then it's harder to track the status and retry failures. 

In dsub, we're built some basic support for batches of like tasks, using a TSV with one row per input parameter set. In CWL, task arrays are handled with dot products and inner products.

This can probably be handled client-side. It's worth discussing though.

Note that this is different from the task list that TES/Funnel currently supports, where a sequence of tasks is executed on a single VM in series. Maybe we could call it task arrays (parallel) and task lists (sequential) or task sequences.

**Comment by buchanae** [comment]
My gut reaction is that client-side is the best place to handle templating, because it's easier to be flexible with the approach to templating there.

Can you give a little more detail on why it's harder to track status and retry? Does this have to do with the client process failing and restarting? Maybe the core issue is that querying for task status (in bulk) isn't easy enough.

**Comment by jbingham** [comment]
Yeah, client side is where we're starting.

Status is harder because you have a large batch. Eg, you submit 100 tasks in parallel to do a parameter sweep. You can list 100, but what you'd really like is a summary, like 50 are pending, 35 are running, 13 succeeded and 2 failed. 

Then it's nice to have a way to resubmit the batch and rerun only the failures. Also, easy to do client-side.

Server-side gets to be useful when you have 100,000 tasks and you don't want to have to stream back such a long list to the client in order to summarize status. 

Or you want to submit 100,000 tasks and know that all are queued. Doing that many HTTP API calls gets to be a bit much for overhead, and you may have a few that fail to get submitted if the server is flaky. If you can instead make one API call, and have a transactional guarantee that everything is enqueued, you can have confidence that the server will handle it.

These are really only a benefit when you've got a lot of tasks. For typical bioinformatics use cases, client-side is fine. I'd expect that at some point in the future, more people will start asking for server-side though. There's a reason that SGE and other job schedulers make task arrays a core feature.

**Comment by buchanae** [comment]
In the case of large batches, I think the best way forward is to add a `CreateTaskBatch` endpoint to TES. What do you think?

```proto
message CreateTaskBatchRequest {
  repeated CreateTaskRequest batch = 1;
}

message TaskBatch {
  repeated Task tasks = 2;
}

service TaskService {
  rpc CreateTaskBatch(CreateTaskBatchRequest) (TaskBatch)
}
```

**Comment by buchanae** [comment]
@jbingham Do you have people creating 100K tasks at once? 

I agree status is more tricky in TES. Filtering the ListTasks() (aka ListJobs) endpoint would be useful here, but hasn't been discussed much nor exercised in Funnel. I think tagging tasks and searching on tags would give TES clients what they need to list and summarize specific task batches. Perhaps CreateTaskBatch could create a tag automatically. I think building the summary is best left to clients than servers.

**Comment by jbingham** [comment]
Not 100k tasks, but 1000. We currently do this client side. The users create a TSV file of input parameters, one row per task. Then they run a single command-line, which creates 1 task per row in the TSV file. 

Then, exactly as you suggest, we add some tags to the tasks so that the CLI can get friendlier status reports, and can cancel all tasks as a batch. This is done client-side. It would be nice to filter the response server-side so it doesn't come back too big for the client to parse. But certainly initially it's totally fine to do client-side.

**Comment by buchanae** [comment]
These issue from TES are related I think:
https://github.com/ga4gh/task-execution-schemas/issues/55
https://github.com/ga4gh/task-execution-schemas/issues/57
https://github.com/ga4gh/task-execution-schemas/issues/60
